lightning/src/ln/features.rs

```diff
@@ -18,17 +18,106 @@ use std::marker::PhantomData;
 use ln::msgs::DecodeError;
 use util::ser::{Readable, Writeable, Writer};
 
-#[macro_use]
-mod sealed { // You should just use the type aliases instead.
-	pub struct InitContext {}
-	pub struct NodeContext {}
-	pub struct ChannelContext {}
-
-	/// An internal trait capturing the various feature context types
-	pub trait Context {}
-	impl Context for InitContext {}
-	impl Context for NodeContext {}
-	impl Context for ChannelContext {}
+mod sealed {
+	/// The context in which [`Features`] are applicable. Defines which features are supported.
+	///
+	/// [`Features`]: ../struct.Features.html
+	pub trait Context {
+		/// Features that are supported by the implementation, indicated by setting their
+		/// optional (odd) bits.
+		const SUPPORTED_FEATURE_FLAGS: &'static [u8];
+
+		/// Bitmask for selecting features that are known though not necessarily implemented.
+		const KNOWN_FEATURE_MASK: &'static [u8];
+
+		/// Bitmask for selecting features that are unknown to the implementation.
+		const UNKNOWN_FEATURE_MASK: &'static [u8];
+	}
+
+	/// Defines a [`Context`] by stating which features it does and does not support. Features are
+	/// specified as a comma-separated list of bytes where each byte is a pipe-delimited list of
+	/// feature identifiers.
+	///
+	/// [`Context`]: trait.Context.html
+	macro_rules! define_context {
+		($context: ident {
+			supported_features: [$( $( $supported_feature: ident )|*, )*],
+			unsupported_features: [$( $( $unsupported_feature: ident )|*, )*],
+		}) => {
+			pub struct $context {}
+
+			impl Context for $context {
+				const SUPPORTED_FEATURE_FLAGS: &'static [u8] = &[
+					$(
+						0b00_00_00_00 $(| <Self as $supported_feature>::OPTIONAL_MASK)*,
+					)*
+				];
+
+				const KNOWN_FEATURE_MASK: &'static [u8] = &[
+					$(
+						0b00_00_00_00 $(|
+							<Self as $supported_feature>::REQUIRED_MASK |
+							<Self as $supported_feature>::OPTIONAL_MASK)*
+						$(|
+							<Self as $unsupported_feature>::REQUIRED_MASK |
+							<Self as $unsupported_feature>::OPTIONAL_MASK)*,
+					)*
+				];
+
+				const UNKNOWN_FEATURE_MASK: &'static [u8] = &[
+					$(
+						0b11_11_11_11 $(&
+							!<Self as $supported_feature>::REQUIRED_MASK &
+							!<Self as $supported_feature>::OPTIONAL_MASK)*
+						$(&
+							!<Self as $unsupported_feature>::REQUIRED_MASK &
+							!<Self as $unsupported_feature>::OPTIONAL_MASK)*,
+					)*
+				];
+			}
+		};
+	}
+
+	define_context!(InitContext {
+		supported_features: [
+			// Byte 0
+			DataLossProtect | UpfrontShutdownScript,
+			// Byte 1
+			VariableLengthOnion | PaymentSecret,
+			// Byte 2
+			BasicMPP,
+		],
+		unsupported_features: [
```

Comment:
I'm confused why this has to be specified explicitly - itn't anything which is not supported unsupported?

Reply:
This is needed in order to form `KNOWN_FEATURE_MASK` and `UNKNOWN_FEATURE_MASK`. Let me know if you can think of a simpler way to form those constants without explicitly giving the unsupported features.